apps/emqx_resource/src/emqx_resource_buffer_worker.erl

```diff
@@ -1317,13 +1361,18 @@ ack_inflight(InflightTID, Ref, Id, Index) ->
                 1;
             [?INFLIGHT_ITEM(Ref, [?QUERY(_, _, _, _) | _] = Batch, _IsRetriable, _WorkerMRef)] ->
                 length(Batch);
-            _ ->
+            [] ->
                 0
         end,
-    IsAcked = Count > 0,
-    IsAcked andalso ets:update_counter(InflightTID, ?SIZE_REF, {2, -Count, 0, 0}),
-    emqx_resource_metrics:inflight_set(Id, Index, inflight_num_msgs(InflightTID)),
-    IsAcked.
+    IsKnownRef = (Count > 0),
+    case IsKnownRef of
+        true ->
+            ets:update_counter(InflightTID, ?SIZE_REF, {2, -Count, 0, 0}),
```

Comment:
Nit: here we have quite the same pattern as @thalesmg spotted, and the check got a bit separated from the update.
Maybe it is worth considering to hide the whole `ets:update_counter(..)` in an inline fun with a guard.